// react-leaflet-polygon-map/dist/CircleApprox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _NumberPicker = require('react-widgets/lib/NumberPicker');

var _NumberPicker2 = _interopRequireDefault(_NumberPicker);

var _cloneDeep = require('lodash/cloneDeep');

var _cloneDeep2 = _interopRequireDefault(_cloneDeep);

var _noop = require('lodash/noop');

var _noop2 = _interopRequireDefault(_noop);

var _cleanProps = require('./cleanProps');

var _cleanProps2 = _interopRequireDefault(_cleanProps);

var _MapHelpers = require('./MapHelpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CircleApprox = function CircleApprox(props) {
  return _react2.default.createElement(
    'div',
    { className: 'map-circle-approx' },
    'Radius (in miles)',
    _react2.default.createElement(_NumberPicker2.default, {
      onChange: _MapHelpers.radiusChange.bind(undefined, props),
      step: 0.1,
      max: 10,
      defaultValue: 0.100,
      min: 0,
      format: '##.###',
      precision: 2
    }),
    _react2.default.createElement(
      'button',
      {
        type: 'button',
        className: 'btn-primary save btn',
        onClick: props.makeCircle
      },
      'Create Circle'
    ),
    _react2.default.createElement(
      'button',
      {
        type: 'button',
        className: 'btn-primary save btn map-toggle-x',
        onClick: function onClick() {
          var p = (0, _cloneDeep2.default)(props);
          p.makeCircleOn = false;
          (0, _cleanProps2.default)(p, props.onShapeChange, _noop2.default);
        }
      },
      'x'
    )
  );
};

exports.default = CircleApprox;


CircleApprox.propTypes = {
  radiusChange: _propTypes2.default.func,
  sidesChange: _propTypes2.default.func,
  makeCircle: _propTypes2.default.func,
  turnOff: _propTypes2.default.func
};
