// react-leaflet-polygon-map/dist/EditTools.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactLeaflet = require('react-leaflet');

var _reactLeafletDraw = require('react-leaflet-draw');

var _cloneDeep = require('lodash/cloneDeep');

var _cloneDeep2 = _interopRequireDefault(_cloneDeep);

var _noop = require('lodash/noop');

var _noop2 = _interopRequireDefault(_noop);

var _updateShapes = require('./updateShapes');

var _updateShapes2 = _interopRequireDefault(_updateShapes);

var _cleanProps = require('./cleanProps');

var _cleanProps2 = _interopRequireDefault(_cleanProps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var EditTools = function EditTools(p) {
  if (p.edit && p.bindPoint !== 'm') {
    return _react2.default.createElement(
      _reactLeaflet.FeatureGroup,
      null,
      _react2.default.createElement(_reactLeafletDraw.EditControl, {
        position: 'topright',
        draw: {
          polyline: false,
          circlemarker: false,
          circle: false,
          marker: {
            icon: p.markerIcon
          }
        },
        edit: {
          edit: false
        },
        onCreated: _updateShapes2.default.bind(undefined, p, p.bindPoint.leafletElement)
      })
    );
  }
  return null;
};

exports.default = EditTools;
